packages/flutter/lib/src/painting/shader_warm_up.dart

```diff
@@ -0,0 +1,144 @@
+    final Canvas canvas = Canvas(recorder);
+
+    warmUpOnCanvas(canvas);
+
+    final Picture picture = recorder.endRecording();
+    final TimelineTask shaderWarmUpTask = TimelineTask();
+    shaderWarmUpTask.start('Warm-up shader');
+    picture.toImage(width, height).then((Image image) {
+      shaderWarmUpTask.finish();
+    });
+  }
+}
+
+/// Default way of warming up Skia shader compilations.
+///
+/// The draw operations being warmed up here are decided according to Flutter
+/// engineers' observation and experience based on the apps and the performance
+/// issues seen so far.
+///
+/// To decide which draw operations to be added to your custom warm up process,
+/// try capture an skp using `flutter screenshot --observatory- port=<port>
+/// --type=skia` and analyze it with https://debugger.skia.org. Alternatively,
+/// one may run the app with `flutter run --trace-skia` and then examine the GPU
+/// thread in the observatory timeline to see which Skia draw operations are
+/// commonly used, and which shader compilations are causing janks.
+///
+/// The size of [width]=[height]=1000 here are picked arbitrarily. The only goal
+/// is to have them not too far away from the actual resolution (very few mobile
+/// devices seem to have a resolution that is 10x larger or 10x smaller than
+/// 1000x1000). The size should have little effect on the shader warm-up. A
+/// custom shader warm up can pick another arbitrary number that's close to the
+/// targeted devices.
+class DefaultShaderWarmUp extends ShaderWarmUp {
+  /// Allow [DefaultShaderWarmUp] to be used as the default value of parameters.
+  const DefaultShaderWarmUp();
+
+  @override
+  int get width => 1000;
+
+  @override
+  int get height => 1000;
+
+  /// Trigger common draw operations on a canvas to warm up GPU shader
+  /// compilation cache.
+  @override
+  void warmUpOnCanvas(Canvas canvas) {
+    final RRect rrect = RRect.fromLTRBXY(20, 20, 60, 60, 10, 10);
+    final Path rrectPath = Path()..addRRect(rrect);
+
+    final Path circlePath = Path()..addOval(
+        Rect.fromCircle(center: const Offset(40, 40), radius: 20)
+    );
+
+    // The following path is copied from
+    // https://skia.org/user/api/SkCanvas_Reference#SkCanvas_drawPath
+    final Path path = Path();
+    path.moveTo(20, 60);
+    path.quadraticBezierTo(60, 20, 60, 60);
+    path.close();
+    path.moveTo(60, 20);
+    path.quadraticBezierTo(60, 60, 20, 60);
+
+    final List<Path> paths = <Path>[rrectPath, circlePath, path];
+
+    final List<Paint> paints = <Paint>[
+      Paint()
+        ..isAntiAlias = true
+        ..style = PaintingStyle.fill,
+      Paint()
+        ..isAntiAlias = true
+        ..style = PaintingStyle.stroke
+        ..strokeWidth = 10,
+      Paint()
+        ..isAntiAlias = true
+        ..style = PaintingStyle.stroke
+        ..strokeWidth = 0.1  // hairline
+    ];
+
+    // Warm up path stroke and fill shaders.
+    for (int i = 0; i < paths.length; i++) {
```

Comment:
fwiw my preference is `i += 1` rather than `i++`, it avoids the confusion about whether the compiler is going to know to treat it as `++i` and discard the value it would otherwise keep around redundantly.